### PR TITLE
KillTask API

### PR DESCRIPTION
**Pull request by ananaysingh**
This pull request contains the code changes and additions made to introduce the feature to kill a specific task of an aurora job

### CODE REVIEW

File: cmd/killTask.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/**
```
Comment (lenhattan86) [inline cmd/killTask.go:1]: We dont need this file, please add a sub command for kill task in kill command. 
So, most of the code must be moved to kill.go

File: cmd/killTask.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	/* Sub-Commands */
+
+	// Kill Job
+	killTask.AddCommand(killTaskCmd)
+
+	killTaskCmd.Flags().StringVarP(env, "environment", "e", "", "Aurora Environment")
+	killTaskCmd.Flags().StringVarP(role, "role", "r", "", "Aurora Role")
+	killTaskCmd.Flags().StringVarP(name, "name", "n", "", "Aurora Name")
+	killTaskCmd.Flags().StringVarP(instance, "instance", "i", "", "Instance number")
```
Comment (lenhattan86) [inline cmd/killTask.go:34]: instance number must be slice []int32 not string.
We can kill a list of instances in the same job.

**Review by lenhattan86 — CHANGES_REQUESTED** [review]

**Comment by lenhattan86** [comment]
also please make change on job.go for GPU(j.GPU).

File: cmd/kill.go [hunk 3]
```diff
@@ -47,6 +61,12 @@ var killJobCmd = &cobra.Command{
 	Run:   killJob,
 }
 
+var killTaskCmd = &cobra.Command{
```
Comment (lenhattan86) [inline cmd/kill.go:64]: var killTasksCmd = &cobra.Command{
	Use:   "tasks",
	Short: "Kill Aurora Tasks",
	Run:   killTasks,
}

File: cmd/kill.go [hunk 4]
```diff
@@ -34,6 +37,17 @@ func init() {
 	killJobCmd.MarkFlagRequired("environment")
 	killJobCmd.MarkFlagRequired("role")
 	killJobCmd.MarkFlagRequired("name")
+
+	//Set flags for killTask sub-command
+	killTaskCmd.Flags().StringVarP(env, "environment", "e", "", "Aurora Environment")
+	killTaskCmd.Flags().StringVarP(role, "role", "r", "", "Aurora Role")
+	killTaskCmd.Flags().StringVarP(name, "name", "n", "", "Aurora Name")
+	killTaskCmd.Flags().StringVarP(instance, "instance", "i", "", "Instance Number")
```
Comment (lenhattan86) [inline cmd/kill.go:45]: lets pass multiple instances to this command.

File: cmd/kill.go [hunk 5]
```diff
@@ -64,3 +84,36 @@ func killJob(cmd *cobra.Command, args []string) {
 		}
 	}
 }
+
+func killTask(cmd *cobra.Command, args []string) {
```
Comment (lenhattan86) [inline cmd/kill.go:88]: killTasks(

**Review by lenhattan86 — CHANGES_REQUESTED** [review]

File: cmd/kill.go [hunk 6]
```diff
@@ -34,6 +38,17 @@ func init() {
 	killJobCmd.MarkFlagRequired("environment")
 	killJobCmd.MarkFlagRequired("role")
 	killJobCmd.MarkFlagRequired("name")
+
+	//Set flags for killTask sub-command
+	killTasksCmd.Flags().StringVarP(env, "environment", "e", "", "Aurora Environment")
+	killTasksCmd.Flags().StringVarP(role, "role", "r", "", "Aurora Role")
+	killTasksCmd.Flags().StringVarP(name, "name", "n", "", "Aurora Name")
+	killTasksCmd.Flags().StringVarP(instance, "instance", "i", "", "Instance Number")
```
Comment (lenhattan86) [inline cmd/kill.go:46]: killTasksCmd.Flags().StringVarP(instances, "instances", "i", "", "Instances")

File: cmd/kill.go [hunk 7]
```diff
@@ -64,3 +85,50 @@ func killJob(cmd *cobra.Command, args []string) {
...
+		if intErr != nil {
+			log.Fatalln("Instance passed should be a number. Error: " + intErr.Error())
+			break
+		} else {
+			instanceList[i] = int32(instanceNumber)
+		}
+	}
+
+	//Call the killtasks function, passing the instanceList as the list of instances to be killed.
+	_, err := client.KillInstances(task.JobKey(), instanceList...)
```
Comment (lenhattan86) [inline cmd/kill.go:124]: if _, err := client.KillInstances(task.JobKey(), instanceList...); err != nil {
	log.Fatalln(err)
}

File: cmd/kill.go [hunk 8]
```diff
@@ -64,3 +85,50 @@ func killJob(cmd *cobra.Command, args []string) {
...
+
+	if err != nil {
+		log.Fatalln(err)
+	}
+	if monitor {
```
Comment (lenhattan86) [inline cmd/kill.go:129]: add a new line before if

File: cmd/kill.go [hunk 9]
```diff
@@ -64,3 +85,50 @@ func killJob(cmd *cobra.Command, args []string) {
...
+	if instance == nil {
+		log.Fatalln("Instance number not found. Please pass a valid instance number. If you want to pass multiple instances, please pass them as space separated integer values")
+	}
+
+	/*
+	* In the following block, we convert instance numbers, which were passed as strings, to integer values
+	* After converting them to integers, we add them to a slice of type int32.
+	 */
+	var intErr error
+	var instanceNumber int
```
Comment (lenhattan86) [inline cmd/kill.go:108]: We don't need line 107 & 108.

File: cmd/kill.go [hunk 10]
```diff
@@ -64,3 +85,50 @@ func killJob(cmd *cobra.Command, args []string) {
...
+
+	splitString := strings.Split(*instance, " ")
```
Comment (lenhattan86) [inline cmd/kill.go:110]: I would prefer instance list like "1,2,5" and it can also be "1, 2, 5"

File: cmd/kill.go [hunk 11]
```diff
@@ -47,6 +62,12 @@ var killJobCmd = &cobra.Command{
 	Run:   killJob,
 }
 
+var killTasksCmd = &cobra.Command{
```
Comment (lenhattan86) [inline cmd/kill.go:65]: please add more description how to use the command.

File: cmd/root.go [hunk 12]
```diff
@@ -51,6 +51,7 @@ var monitor bool
 var timeout time.Duration
 var log = logrus.New()
 var taskStatus = new(string)
+var instance = new(string)
```
Comment (lenhattan86) [inline cmd/root.go:54]: do we still need this?

**Review by lenhattan86 — CHANGES_REQUESTED** [review]

File: cmd/root.go (hunk 12 above)
Comment (ananaysingh) [reply cmd/root.go:54]: Yes we need it. This is the variable in which we capture, as string, the passed instanceNumbers. All the flags, used by other commands as well, are defined here. So I felt it is apt to define this flag here

File: cmd/kill.go (hunk 9 above)
Comment (ananaysingh) [reply cmd/kill.go:108]: We need the two variables for the loop on line 113. I tried without them, golang won't allow

File: cmd/kill.go (hunk 8 above)
Comment (ananaysingh) [reply cmd/kill.go:141]: Added

File: cmd/kill.go (hunk 11 above)
Comment (ananaysingh) [reply cmd/kill.go:77]: Added description

File: cmd/kill.go (hunk 9 above)
Comment (lenhattan86) [reply cmd/kill.go:120]: in line 127, you can do 

instanceNumber, intErr := strconv.Atoi(splitString[i]) right?

File: cmd/kill.go [hunk 13]
```diff
@@ -34,6 +38,17 @@ func init() {
 	killJobCmd.MarkFlagRequired("environment")
 	killJobCmd.MarkFlagRequired("role")
 	killJobCmd.MarkFlagRequired("name")
+
+	//Set flags for killTask sub-command
+	killTasksCmd.Flags().StringVarP(env, "environment", "e", "", "Aurora Environment")
+	killTasksCmd.Flags().StringVarP(role, "role", "r", "", "Aurora Role")
+	killTasksCmd.Flags().StringVarP(name, "name", "n", "", "Aurora Name")
+	killTasksCmd.Flags().StringVarP(instances, "instances", "i", "", "Instances")
```
Comment (lenhattan86) [inline cmd/kill.go:46]: killTasksCmd.Flags().StringVarP(instances, "instances", "i", "", "Instances e.g. 1, 2, 5")

**Review by lenhattan86 — CHANGES_REQUESTED** [review]

File: cmd/kill.go [hunk 14]
```diff
@@ -64,3 +97,50 @@ func killJob(cmd *cobra.Command, args []string) {
...
+
+	splitString := strings.Split(*instances, ",")
+	instanceList := make([]int32, len(splitString))
+
+	for i := range instanceList {
+		splitString[i] = strings.TrimSpace(splitString[i])
+		instanceNumber, intErr = strconv.Atoi(splitString[i])
+		if intErr != nil {
+			log.Fatalln("Instance passed should be a number. Error: " + intErr.Error())
+			break
```
Comment (lenhattan86) [inline cmd/kill.go:130]: I don't think we need `break` here but just in case we should put return instead.

**Review by lenhattan86 — CHANGES_REQUESTED** [review]

File: cmd/kill.go (hunk 9 above)
Comment (ananaysingh) [reply cmd/kill.go:120]: Nope. Apparently golang requires variables to be used in a loop to be defined before the loop. I tried to do it the way you want, but it throws compilation errors

**Review by lenhattan86 — APPROVED** [review]

**Pull request merged**